src/pybind/mgr/cephadm/module.py

```diff
@@ -1693,8 +1694,9 @@ def update_mons(self, spec):
             # Improve Error message. Point to parse_host_spec examples
             raise orchestrator.OrchestratorValidationError("Mons need a host spec. (host, network, name(opt))")
 
-        spec = NodeAssignment(spec=spec, get_hosts_func=self._get_hosts, service_type='mon').load()
-        return self._update_mons(spec)
+        assigned_spec = NodeAssignment(spec=spec, get_hosts_func=self._get_hosts, service_type='mon').load()
+        assert isinstance(assigned_spec, orchestrator.StatefulServiceSpec)
```

Comment:
```
2020-02-04T17:49:58.192 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: debug 2020-02-04T17:49:58.191+0000 7f74e470a700 -1 Remote method threw exception: Traceback (most recent call last):
2020-02-04T17:49:58.192 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: File "/usr/share/ceph/mgr/cephadm/module.py", line 1700, in update_mons
2020-02-04T17:49:58.193 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: assert isinstance(assigned_spec, orchestrator.StatefulServiceSpec)
2020-02-04T17:49:58.193 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: AssertionError
2020-02-04T17:49:58.193 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: debug 2020-02-04T17:49:58.191+0000 7f74e470a700 -1 mgr handle_command module 'orchestrator_cli' command handler threw exception: Remote method threw exception: Traceback (most recent call last):
2020-02-04T17:49:58.194 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: File "/usr/share/ceph/mgr/cephadm/module.py", line 1700, in update_mons
2020-02-04T17:49:58.194 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: assert isinstance(assigned_spec, orchestrator.StatefulServiceSpec)
2020-02-04T17:49:58.195 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: AssertionError
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:Error EINVAL: Traceback (most recent call last):
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1069, in _handle_command
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:    return CLICommand.COMMANDS[cmd['prefix']].call(self, cmd, inbuf)
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 309, in call
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:    return self.func(mgr, **kwargs)
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 140, in wrapper
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:    return func(*args, **kwargs)
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/orchestrator_cli/module.py", line 641, in _update_mons
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:    completion = self.update_mons(spec)
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 1445, in inner
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:    completion = self._oremote(method_name, args, kwargs)
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 1515, in _oremote
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:    return mgr.remote(o, meth, *args, **kwargs)
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1429, in remote
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:    args, kwargs)
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:RuntimeError: Remote method threw exception: Traceback (most recent call last):
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/cephadm/module.py", line 1700, in update_mons
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:    assert isinstance(assigned_spec, orchestrator.StatefulServiceSpec)
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:AssertionError
```
/a/sage-2020-02-04_16:34:01-rados-wip-sage-testing-2020-02-04-0730-distro-basic-smithi/4732250

Reply:
Yay! this looks like a real thing! I always assumed that we're getting a StatefulServiceSpec

Reply:
#33175